shtdlib.sh

```diff
@@ -1996,13 +1996,44 @@ function ln_sf {
 # Create string of random characters
 #  - First param is length, default: 20
 #  - Second param is characters, default: A-Za-z0-9_ (Note: '-' specifies range)
-gen_rand_chars() {
+function gen_rand_chars {
+    shopt_decorator_option_name='pipefail'
+    shopt_decorator_option_value='false'
+    # shellcheck disable=2015
+    shopt_decorator "${FUNCNAME[0]}" "${@:-}" && return || conditional_exit_on_fail 121 "Failed to run ${FUNCNAME[0]} with shopt_decorator"
```

Comment:
Shouldn't the exit code be 141 instead of 121?

Reply:
I'm actually not sure where the code comes from now that I think about it,  docs I find say 13, are you getting a 141?

I'm wondering if it's different between platforms?

Reply:
which number are you getting for SIGPIPE?

Reply:
It returns 141.   The best explanation I could find is here: http://blogs.escenic.com/rd/2017/A-tale-of-broken-pipes

Relevant section:

The return code will be 0 because head -n 1 can probably never fail, but the PIPESTATUS will reveal that both find and grep failed with the return code 141.

In bash(1) (the man-page), we can see that an exit code for a process is its return code, or 128+n if it is terminated with a signal. 141 is 13 more than 128, and indeed, according to signal(7), Signal 13 is SIGPIPE:

```SIGPIPE      13       Term    Broken pipe: write to pipe with no readers```

This means that in any pipeline of commands, the downstream of a pipe can close the pipe, and cause upstream ripple effects. They will try to write to the pipe and will abruptly be killed with SIG_PIPE.

Reply:
I've also been using 141 in my branch to prevent this error, so I believe it's correct at least on Linux-based platforms.

Reply:
Makes sense, I'll change to 141, I think I remember reading that 128+n before so 121 must be a typo or something.

Reply:
Err, I re-read the code and realized what's up and why it's 121.

121 is the exit code returned by the decorator to indicate that it's already run the decorated command and that no more recursion will happen.


` 211 function shopt_decorator {                                                      
 212     debug 10 "${FUNCNAME} used with ${*}"                                       
 213     if [ -n "${shopt_decorator_option_value:-}" ]  && [ -n "$(shopt -o "${shopt_decorator_option_name:-}")" ] ; then
 214         if [ "${FUNCNAME[0]}" != "${FUNCNAME[2]}" ] ; then                      
 215             if shopt -qo "${shopt_decorator_option_name}" ; then                
 216                 # Option is set                                                 
 217                 if ! "${shopt_decorator_option_value}" ; then                   
 218                     # Option should not be set                                  
 219                     debug 10 "Temporarily unsetting bash option ${shopt_decorator_option_name}"
 220                     shopt -uo "${shopt_decorator_option_name}"                  
 221                 else                                                            
 222                     debug 10 "No need to set ${shopt_decorator_option_name}, it's already ${shopt_decorator_option_value}"
 223                 fi                                                              
 224                 "${@}"                                                          
 225                 return_code="${?}"                                              
 226                 debug 10 "Got return code ${return_code}"                       
 227                 # Set the option again in case it was unset                     
 228                 debug 10 "(Re)Setting ${shopt_decorator_option_name}"           
 229                 shopt -so "${shopt_decorator_option_name}"                      
 230                 return ${return_code}                                           
 231             else                                                                
 232                 # Option is not set                                             
 233                 if "${shopt_decorator_option_value}" ; then                     
 234                     # Option should be set                                      
 235                     debug 10 "Temporarily setting bash option ${shopt_decorator_option_name}"
 236                     shopt -so "${shopt_decorator_option_name}"                  
 237                 else                                                            
 238                     debug 10 "No need to unset ${shopt_decorator_option_name}, it's already ${shopt_decorator_option_value}"
 239                 fi                                                              
 240                 "${@}"                                                          
 241                 return_code="${?}"                                              
 242                 debug 10 "Got return code ${return_code}"                       
 243                 # Unset the option in case it was set                           
 244                 debug 10 "(Re)Unsetting ${shopt_decorator_option_name}"         
 245                 shopt -uo "${shopt_decorator_option_name}"                      
 246                 return ${return_code}                                           
 247             fi                                                                  
 248         fi                                                                      
 249         # Calling function is the decorator, skip                               
 250         return 121                                                              
 251     else                                                                        
 252         color_echo red "Called ${FUNCNAME[*]} without setting required variables with valid option name/value. The variables shopt_decorator_option_name and shopt_decorator_option_value need to be set to a valid shopt option and a command/function that evaluates true/false, 'true'/'false' are valid commands"
 253         exit 126                                                                
 254     fi                                                                          
 255     # We should never get here                                                  
 256     exit 127                                                                    
 257 }`